jep/0000/README.adoc

```diff
@@ -0,0 +1,220 @@
+//| :bulb: JEP-NUMBER :bulb:
+//
+//
+// Uncomment when this JEP status is set to Accepted, Rejected or Withdrawn.
+//| Resolution
+//| :bulb: Link to relevant post in the jenkinsci-dev@ mailing list archives :bulb:
+
+|===
+
+
+== Abstract
+
+Configuration-as-Code (JEP-201) do describe the mechanisms used to configure a Jenkins master based
+on yaml file(s), by discovering the data model from Java API, without the need to write any "glue-code"
+for each and every supported plugin.
+
+This approach comes with some drawback: Any update to Jenkins master's core or plugins will change the
+inferred model, and as such a yaml configuration used a t time might not be valid a t+1. This concern
+has been raised by many early-adopters.
+
+This proposal is about defining the adequate tooling to balance this limitation, so both end-users
+and plugin maintainer are able to detect such compatibility breaks, and provide some diagnostic and
+resolution hints.
+
+== Specification
+
+Configuration-as-Code tooling is designed to fill the gap between the raw mechanism defined by JEP-201
+to configure a specific jenkins-core + plugins combination, and the need to keep this configuration
+stable during core or plugin upgrades, or at least detect breaking changes and provide hints to manage
+upgreade.
+
+This includes:
+
+- tooling to validate configuration, sort of a "dry-run" mode, with adequate diagnostic when some
+configuration mismatch is detected.
+- tooling to compare inferred model between versions, to be used by plugin developers to be aware
+changes they made during development might have impact on Configuration-as-Code.
+- Document recommendations for plugin developers on ways to avoid breaking changes
+- Metadata definition for plugin developers to document breaking changes and how to fix them. Based
+on this, when yaml validation do detect a configuration mismatch it can offer migration hints.
+
+
+== Motivation
+
+Configuration-as-Code doesn't define a data model to setup jenkins, it's based on inferring a model
+from live jenkins instance from it's Java API, especially UI data-binding mechanisms. Based on this,
+changes made to core or a plugin _might_ change this model and a configuration file _could_ not be
+valid after upgrade.
+
+There's not so much cases where such compatibility breaks could happen, but there's chances this would.
+Compared to other tools, like `docker-compose`, we can't provide a `version` attribute and accept
+distinct models, as we only know about the live, inferred one.
+
+Configuration-as-Code tooling is defined to workaround this limitation by providing a toolset to
+discover such breaking changes early, before the actual production configuration is updated.
+It will help detect breaking changes during jenkins and plugins developement, offer an option for
+plugin developers to document required breaking changes, and validate a configuration for a specific
+jenkins instance before it get actually applied.
+
+== Reasoning
```

Comment:
Something not addressed in [JEP-201 itself](https://github.com/jenkinsci/jep/blob/master/jep/201/README.adoc#general-mechanism) is why you chose to use databinding (~ `DescribableModel`) as the source of truth, when plugins have traditionally maintained settings compatibility via their XStream form (~ fields, `readResolve`)—we _do_ presume already that a `config.xml` which worked in an older “version” of Jenkins can be loaded safely in a newer version. Thus JEP-201 is introducing an _additional_ compatibility constraint: plugin developers need to be aware of, and work around, both models when refactoring settings.

(Pipeline was already imposing a similar constraint for `Step`s and those `Describable`s being used inside `Step`s. OTOH it also allows `StepDescriptor.newInstance` / `.uninstantiate` to be overridden to customize the process, as I in fact did recently in https://github.com/jenkinsci/docker-workflow-plugin/pull/138. This is not available when sharing code with freestyle via `CoreStep` / `CoreWrapperStep`. I think `job-dsl` also introduced the ability to introspect stuff and pay attention to `@Symbol` but AFAIK most bindings are still hand-written.)